{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1c5a0d4-c952-4ea5-b552-44dfd23ae53d",
   "metadata": {},
   "source": [
    "# Association between host gene expression and behavior\n",
    "Assess to what extent gene expression in the adrenal glands (Adr) and visceral adipose tissue (VAT) predict host behavior."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb9bbced-4773-4fc5-a570-02cc32e7426e",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc603815-5690-48e4-92ed-01cba0718ffc",
   "metadata": {},
   "source": [
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c5bd0635-8c73-4e21-823b-75c310285c7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import linregress\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.model_selection import KFold, cross_val_predict, permutation_test_score\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "matplotlib.use('Agg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f76fbac-b64e-474e-8016-21fe47389cb9",
   "metadata": {},
   "source": [
    "### Set up argument parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "289d1586-f732-4287-b9de-550b12b7d5a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] --output_dir OUTPUT_DIR\n",
      "                             [--phenotype_vars PHENOTYPE_VARS [PHENOTYPE_VARS ...]]\n",
      "                             --phenotype_file PHENOTYPE_FILE --tissue\n",
      "                             {Adr,VAT} --transcriptome_file TRANSCRIPTOME_FILE\n",
      "ipykernel_launcher.py: error: the following arguments are required: --output_dir, --phenotype_file, --tissue, --transcriptome_file\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    }
   ],
   "source": [
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser(description=\"Run Random Forest to predict behavior from transcriptome data.\")\n",
    "    parser.add_argument(\"--output_dir\", required=True, help=\"Directory to save output files\")\n",
    "    parser.add_argument(\"--phenotype_vars\", type=str, nargs=\"+\", help=\"Behavioral variables\")\n",
    "    parser.add_argument(\"--phenotype_file\", type=str, required=True, help=\"Path to Behavioral data file\")\n",
    "    parser.add_argument(\"--tissue\", type=str, required=True, choices=[\"Adr\", \"VAT\"], help=\"Tissue type for transcriptome data (Adr or VAT)\")\n",
    "    parser.add_argument(\"--transcriptome_file\", type=str, required=True, help=\"Path to transcriptome data file (Adr or VAT)\")\n",
    "    return parser.parse_args()\n",
    "\n",
    "# Parse arguments\n",
    "args = parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c63553c5-4717-452a-947f-bb451f5fd966",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulate arguments for troubleshooting\n",
    "\n",
    "class Args:\n",
    "    output_dir = \"output/Adr\"\n",
    "    phenotype_vars = [\"Center_occupancy\", \"Grooming_duration\", \"Social_preference\"]\n",
    "    phenotype_file = \"Behavior_data.txt\"\n",
    "    tissue = \"Adr\"\n",
    "    transcriptome_file = \"Adr_normalized_counts.txt\"\n",
    "\n",
    "args = Args()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e042569b-ca24-4be5-acd4-b05d0e2b7535",
   "metadata": {},
   "source": [
    "### Set up output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fd766bfa-efdf-4698-bef5-f21e65235b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.path.join(args.output_dir, args.tissue)\n",
    "os.makedirs(output_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e335626-03aa-45a2-b740-66eae18cf08b",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9335afde-ee78-45b0-8f87-af87ee739cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load phenotype metadata\n",
    "phenotype = pd.read_csv(args.phenotype_file, sep=\"\\t\")\n",
    "\n",
    "# Load selected tissue's transcriptome data\n",
    "transcriptome = pd.read_csv(args.transcriptome_file, sep=\"\\t\", index_col=0).T.reset_index().rename(columns={\"index\": \"Mouse_ID\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a3be8fd8-d655-45ad-88bb-8200920c12db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>DOB</th>\n",
       "      <th>Dame</th>\n",
       "      <th>Sire</th>\n",
       "      <th>Parents</th>\n",
       "      <th>Litter</th>\n",
       "      <th>Weaned</th>\n",
       "      <th>Experiement_Start</th>\n",
       "      <th>...</th>\n",
       "      <th>TMT</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Pair_mouse</th>\n",
       "      <th>Cage_ID</th>\n",
       "      <th>Treatment_Timepoint</th>\n",
       "      <th>Stress</th>\n",
       "      <th>Center_occupancy</th>\n",
       "      <th>Grooming_duration</th>\n",
       "      <th>Social_preference</th>\n",
       "      <th>Weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2028</td>\n",
       "      <td>2028_T5</td>\n",
       "      <td>Female</td>\n",
       "      <td>01/01/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA_NA</td>\n",
       "      <td>NA_NA_44562</td>\n",
       "      <td>31/01/2022</td>\n",
       "      <td>15/02/2022</td>\n",
       "      <td>...</td>\n",
       "      <td>H2O</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>2029.0</td>\n",
       "      <td>2028_2029</td>\n",
       "      <td>Pair_H2O_T5</td>\n",
       "      <td>Control</td>\n",
       "      <td>344.533333</td>\n",
       "      <td>1</td>\n",
       "      <td>0.093117</td>\n",
       "      <td>15.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2029</td>\n",
       "      <td>2029_T5</td>\n",
       "      <td>Female</td>\n",
       "      <td>01/01/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA_NA</td>\n",
       "      <td>NA_NA_44562</td>\n",
       "      <td>31/01/2022</td>\n",
       "      <td>15/02/2022</td>\n",
       "      <td>...</td>\n",
       "      <td>H2O</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>2028.0</td>\n",
       "      <td>2028_2029</td>\n",
       "      <td>Pair_H2O_T5</td>\n",
       "      <td>Control</td>\n",
       "      <td>121.866667</td>\n",
       "      <td>65</td>\n",
       "      <td>0.446575</td>\n",
       "      <td>16.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>594</td>\n",
       "      <td>594_T5</td>\n",
       "      <td>Male</td>\n",
       "      <td>01/01/2022</td>\n",
       "      <td>550</td>\n",
       "      <td>560.0</td>\n",
       "      <td>550_560</td>\n",
       "      <td>550_560_44562</td>\n",
       "      <td>24/01/2022</td>\n",
       "      <td>15/02/2022</td>\n",
       "      <td>...</td>\n",
       "      <td>H2O</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>595.0</td>\n",
       "      <td>594_595</td>\n",
       "      <td>Pair_H2O_T5</td>\n",
       "      <td>Control</td>\n",
       "      <td>195.866667</td>\n",
       "      <td>14</td>\n",
       "      <td>0.895172</td>\n",
       "      <td>20.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>595</td>\n",
       "      <td>595_T5</td>\n",
       "      <td>Male</td>\n",
       "      <td>01/01/2022</td>\n",
       "      <td>550</td>\n",
       "      <td>560.0</td>\n",
       "      <td>550_560</td>\n",
       "      <td>550_560_44562</td>\n",
       "      <td>24/01/2022</td>\n",
       "      <td>15/02/2022</td>\n",
       "      <td>...</td>\n",
       "      <td>H2O</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>594.0</td>\n",
       "      <td>594_595</td>\n",
       "      <td>Pair_H2O_T5</td>\n",
       "      <td>Control</td>\n",
       "      <td>201.933333</td>\n",
       "      <td>88</td>\n",
       "      <td>0.474453</td>\n",
       "      <td>20.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>596</td>\n",
       "      <td>596_T5</td>\n",
       "      <td>Female</td>\n",
       "      <td>01/01/2022</td>\n",
       "      <td>550</td>\n",
       "      <td>560.0</td>\n",
       "      <td>550_560</td>\n",
       "      <td>550_560_44562</td>\n",
       "      <td>24/01/2022</td>\n",
       "      <td>15/02/2022</td>\n",
       "      <td>...</td>\n",
       "      <td>TMT</td>\n",
       "      <td>Pair_TMT</td>\n",
       "      <td>597.0</td>\n",
       "      <td>596_597</td>\n",
       "      <td>Pair_TMT_T5</td>\n",
       "      <td>Stressor</td>\n",
       "      <td>216.566667</td>\n",
       "      <td>17</td>\n",
       "      <td>0.879760</td>\n",
       "      <td>16.18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mouse_ID Sample_ID     Sex         DOB Dame   Sire  Parents         Litter  \\\n",
       "0      2028   2028_T5  Female  01/01/2022  NaN    NaN    NA_NA    NA_NA_44562   \n",
       "1      2029   2029_T5  Female  01/01/2022  NaN    NaN    NA_NA    NA_NA_44562   \n",
       "2       594    594_T5    Male  01/01/2022  550  560.0  550_560  550_560_44562   \n",
       "3       595    595_T5    Male  01/01/2022  550  560.0  550_560  550_560_44562   \n",
       "4       596    596_T5  Female  01/01/2022  550  560.0  550_560  550_560_44562   \n",
       "\n",
       "       Weaned Experiement_Start  ...  TMT  Treatment Pair_mouse    Cage_ID  \\\n",
       "0  31/01/2022        15/02/2022  ...  H2O   Pair_H2O     2029.0  2028_2029   \n",
       "1  31/01/2022        15/02/2022  ...  H2O   Pair_H2O     2028.0  2028_2029   \n",
       "2  24/01/2022        15/02/2022  ...  H2O   Pair_H2O      595.0    594_595   \n",
       "3  24/01/2022        15/02/2022  ...  H2O   Pair_H2O      594.0    594_595   \n",
       "4  24/01/2022        15/02/2022  ...  TMT   Pair_TMT      597.0    596_597   \n",
       "\n",
       "  Treatment_Timepoint    Stress Center_occupancy Grooming_duration  \\\n",
       "0         Pair_H2O_T5   Control       344.533333                 1   \n",
       "1         Pair_H2O_T5   Control       121.866667                65   \n",
       "2         Pair_H2O_T5   Control       195.866667                14   \n",
       "3         Pair_H2O_T5   Control       201.933333                88   \n",
       "4         Pair_TMT_T5  Stressor       216.566667                17   \n",
       "\n",
       "  Social_preference  Weight  \n",
       "0          0.093117   15.04  \n",
       "1          0.446575   16.09  \n",
       "2          0.895172   20.35  \n",
       "3          0.474453   20.31  \n",
       "4          0.879760   16.18  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Xkr4</th>\n",
       "      <th>Gm53491</th>\n",
       "      <th>Rp1</th>\n",
       "      <th>Sox17</th>\n",
       "      <th>Gm22307</th>\n",
       "      <th>Gm46082</th>\n",
       "      <th>Gm29874</th>\n",
       "      <th>Mrpl15</th>\n",
       "      <th>Lypla1</th>\n",
       "      <th>...</th>\n",
       "      <th>TrnC</th>\n",
       "      <th>COX1</th>\n",
       "      <th>ATP6</th>\n",
       "      <th>ND3</th>\n",
       "      <th>ND4</th>\n",
       "      <th>TrnL2</th>\n",
       "      <th>ND5</th>\n",
       "      <th>ND6</th>\n",
       "      <th>CYTB</th>\n",
       "      <th>TrnP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>594</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>16.416488</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>252.403503</td>\n",
       "      <td>69.770074</td>\n",
       "      <td>...</td>\n",
       "      <td>8.208244</td>\n",
       "      <td>6917.497632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>413379.480243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9870.413411</td>\n",
       "      <td>10525.020870</td>\n",
       "      <td>33748.195209</td>\n",
       "      <td>449.401359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>595</td>\n",
       "      <td>3.275186</td>\n",
       "      <td>3.930223</td>\n",
       "      <td>0.655037</td>\n",
       "      <td>19.651114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.655037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>194.546026</td>\n",
       "      <td>60.263415</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6446.875378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>586866.755805</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12818.421487</td>\n",
       "      <td>24771.538933</td>\n",
       "      <td>98703.614050</td>\n",
       "      <td>126.422165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>596</td>\n",
       "      <td>5.852659</td>\n",
       "      <td>5.164111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.803784</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.857678</td>\n",
       "      <td>61.969329</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344274</td>\n",
       "      <td>8238.822253</td>\n",
       "      <td>0.688548</td>\n",
       "      <td>11.361044</td>\n",
       "      <td>145138.020516</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30920.285112</td>\n",
       "      <td>34343.401976</td>\n",
       "      <td>138186.438931</td>\n",
       "      <td>104.315037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>597</td>\n",
       "      <td>8.006223</td>\n",
       "      <td>3.002334</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.004667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>274.213135</td>\n",
       "      <td>86.066897</td>\n",
       "      <td>...</td>\n",
       "      <td>8.006223</td>\n",
       "      <td>8677.744881</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>218419.769390</td>\n",
       "      <td>1.000778</td>\n",
       "      <td>12487.706220</td>\n",
       "      <td>11304.786781</td>\n",
       "      <td>24252.850820</td>\n",
       "      <td>955.742863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>598</td>\n",
       "      <td>7.682467</td>\n",
       "      <td>2.743738</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.292486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>151.454341</td>\n",
       "      <td>38.961080</td>\n",
       "      <td>...</td>\n",
       "      <td>1.646243</td>\n",
       "      <td>7661.065409</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>635012.382650</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14747.043320</td>\n",
       "      <td>18490.050781</td>\n",
       "      <td>42315.025843</td>\n",
       "      <td>694.165729</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23601 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Mouse_ID      Xkr4   Gm53491       Rp1      Sox17  Gm22307   Gm46082  \\\n",
       "0         594  2.052061  2.052061  2.052061  16.416488      0.0  0.000000   \n",
       "1         595  3.275186  3.930223  0.655037  19.651114      0.0  0.655037   \n",
       "2         596  5.852659  5.164111  0.000000  14.803784      0.0  0.000000   \n",
       "3         597  8.006223  3.002334  0.000000   6.004667      0.0  0.000000   \n",
       "4         598  7.682467  2.743738  0.000000   3.292486      0.0  0.000000   \n",
       "\n",
       "Gene  Gm29874      Mrpl15     Lypla1  ...      TrnC         COX1      ATP6  \\\n",
       "0         0.0  252.403503  69.770074  ...  8.208244  6917.497632  0.000000   \n",
       "1         0.0  194.546026  60.263415  ...  0.000000  6446.875378  0.000000   \n",
       "2         0.0  152.857678  61.969329  ...  0.344274  8238.822253  0.688548   \n",
       "3         0.0  274.213135  86.066897  ...  8.006223  8677.744881  0.000000   \n",
       "4         0.0  151.454341  38.961080  ...  1.646243  7661.065409  0.000000   \n",
       "\n",
       "Gene        ND3            ND4     TrnL2           ND5           ND6  \\\n",
       "0      2.052061  413379.480243  0.000000   9870.413411  10525.020870   \n",
       "1      0.000000  586866.755805  0.000000  12818.421487  24771.538933   \n",
       "2     11.361044  145138.020516  0.000000  30920.285112  34343.401976   \n",
       "3      0.000000  218419.769390  1.000778  12487.706220  11304.786781   \n",
       "4      0.000000  635012.382650  0.000000  14747.043320  18490.050781   \n",
       "\n",
       "Gene           CYTB        TrnP  \n",
       "0      33748.195209  449.401359  \n",
       "1      98703.614050  126.422165  \n",
       "2     138186.438931  104.315037  \n",
       "3      24252.850820  955.742863  \n",
       "4      42315.025843  694.165729  \n",
       "\n",
       "[5 rows x 23601 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 47 entries, 0 to 46\n",
      "Data columns (total 27 columns):\n",
      " #   Column               Non-Null Count  Dtype  \n",
      "---  ------               --------------  -----  \n",
      " 0   Mouse_ID             47 non-null     int64  \n",
      " 1   Sample_ID            47 non-null     object \n",
      " 2   Sex                  47 non-null     object \n",
      " 3   DOB                  47 non-null     object \n",
      " 4   Dame                 45 non-null     object \n",
      " 5   Sire                 45 non-null     float64\n",
      " 6   Parents              47 non-null     object \n",
      " 7   Litter               47 non-null     object \n",
      " 8   Weaned               47 non-null     object \n",
      " 9   Experiement_Start    47 non-null     object \n",
      " 10  Starting_Age         47 non-null     int64  \n",
      " 11  Days_P_Wean          47 non-null     int64  \n",
      " 12  Cohort               47 non-null     object \n",
      " 13  Timepoint            47 non-null     object \n",
      " 14  Sampling_date        47 non-null     object \n",
      " 15  Sampling_Age         47 non-null     int64  \n",
      " 16  Housing              47 non-null     object \n",
      " 17  TMT                  47 non-null     object \n",
      " 18  Treatment            47 non-null     object \n",
      " 19  Pair_mouse           30 non-null     float64\n",
      " 20  Cage_ID              47 non-null     object \n",
      " 21  Treatment_Timepoint  47 non-null     object \n",
      " 22  Stress               47 non-null     object \n",
      " 23  Center_occupancy     47 non-null     float64\n",
      " 24  Grooming_duration    47 non-null     int64  \n",
      " 25  Social_preference    47 non-null     float64\n",
      " 26  Weight               47 non-null     float64\n",
      "dtypes: float64(5), int64(5), object(17)\n",
      "memory usage: 10.0+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 40 entries, 0 to 39\n",
      "Columns: 23601 entries, Mouse_ID to TrnP\n",
      "dtypes: float64(23600), object(1)\n",
      "memory usage: 7.2+ MB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Data overview\n",
    "\n",
    "# Display first 5 rows\n",
    "display(phenotype.head())\n",
    "display(transcriptome.head())\n",
    "\n",
    "# Check data types and missing values\n",
    "print(phenotype.info())\n",
    "print(transcriptome.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "443cf39e-40c3-4c54-ac6e-bb045bfc57bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert transcriptome values to numeric (force non-numeric values to NaN and handle appropriately)\n",
    "transcriptome.iloc[:, 1:] = transcriptome.iloc[:, 1:].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "# Select relevant columns from phenotype metadata\n",
    "phenotype = phenotype[[\"Mouse_ID\", \"Treatment\"] + args.phenotype_vars]  # Dynamic selection of phenotype variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7c55ee49-0d5e-4181-921d-c6cdde60aeb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Center_occupancy</th>\n",
       "      <th>Grooming_duration</th>\n",
       "      <th>Social_preference</th>\n",
       "      <th>Xkr4</th>\n",
       "      <th>Gm53491</th>\n",
       "      <th>Rp1</th>\n",
       "      <th>Sox17</th>\n",
       "      <th>Gm22307</th>\n",
       "      <th>...</th>\n",
       "      <th>TrnC</th>\n",
       "      <th>COX1</th>\n",
       "      <th>ATP6</th>\n",
       "      <th>ND3</th>\n",
       "      <th>ND4</th>\n",
       "      <th>TrnL2</th>\n",
       "      <th>ND5</th>\n",
       "      <th>ND6</th>\n",
       "      <th>CYTB</th>\n",
       "      <th>TrnP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2028</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>344.533333</td>\n",
       "      <td>1</td>\n",
       "      <td>0.093117</td>\n",
       "      <td>3.862889</td>\n",
       "      <td>3.090312</td>\n",
       "      <td>0.772578</td>\n",
       "      <td>12.361246</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.772578</td>\n",
       "      <td>8639.738516</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.317734</td>\n",
       "      <td>362905.328813</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16425.005888</td>\n",
       "      <td>25810.282064</td>\n",
       "      <td>66233.102292</td>\n",
       "      <td>505.265938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>594</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>195.866667</td>\n",
       "      <td>14</td>\n",
       "      <td>0.895172</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>16.416488</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.208244</td>\n",
       "      <td>6917.497632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.052061</td>\n",
       "      <td>413379.480243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9870.413411</td>\n",
       "      <td>10525.020870</td>\n",
       "      <td>33748.195209</td>\n",
       "      <td>449.401359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>595</td>\n",
       "      <td>Pair_H2O</td>\n",
       "      <td>201.933333</td>\n",
       "      <td>88</td>\n",
       "      <td>0.474453</td>\n",
       "      <td>3.275186</td>\n",
       "      <td>3.930223</td>\n",
       "      <td>0.655037</td>\n",
       "      <td>19.651114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6446.875378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>586866.755805</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12818.421487</td>\n",
       "      <td>24771.538933</td>\n",
       "      <td>98703.614050</td>\n",
       "      <td>126.422165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>596</td>\n",
       "      <td>Pair_TMT</td>\n",
       "      <td>216.566667</td>\n",
       "      <td>17</td>\n",
       "      <td>0.879760</td>\n",
       "      <td>5.852659</td>\n",
       "      <td>5.164111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.803784</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344274</td>\n",
       "      <td>8238.822253</td>\n",
       "      <td>0.688548</td>\n",
       "      <td>11.361044</td>\n",
       "      <td>145138.020516</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30920.285112</td>\n",
       "      <td>34343.401976</td>\n",
       "      <td>138186.438931</td>\n",
       "      <td>104.315037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>597</td>\n",
       "      <td>Pair_TMT</td>\n",
       "      <td>118.633333</td>\n",
       "      <td>107</td>\n",
       "      <td>0.380531</td>\n",
       "      <td>8.006223</td>\n",
       "      <td>3.002334</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.004667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.006223</td>\n",
       "      <td>8677.744881</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>218419.769390</td>\n",
       "      <td>1.000778</td>\n",
       "      <td>12487.706220</td>\n",
       "      <td>11304.786781</td>\n",
       "      <td>24252.850820</td>\n",
       "      <td>955.742863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23605 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Mouse_ID Treatment  Center_occupancy  Grooming_duration  Social_preference  \\\n",
       "0     2028  Pair_H2O        344.533333                  1           0.093117   \n",
       "1      594  Pair_H2O        195.866667                 14           0.895172   \n",
       "2      595  Pair_H2O        201.933333                 88           0.474453   \n",
       "3      596  Pair_TMT        216.566667                 17           0.879760   \n",
       "4      597  Pair_TMT        118.633333                107           0.380531   \n",
       "\n",
       "       Xkr4   Gm53491       Rp1      Sox17  Gm22307  ...      TrnC  \\\n",
       "0  3.862889  3.090312  0.772578  12.361246      0.0  ...  0.772578   \n",
       "1  2.052061  2.052061  2.052061  16.416488      0.0  ...  8.208244   \n",
       "2  3.275186  3.930223  0.655037  19.651114      0.0  ...  0.000000   \n",
       "3  5.852659  5.164111  0.000000  14.803784      0.0  ...  0.344274   \n",
       "4  8.006223  3.002334  0.000000   6.004667      0.0  ...  8.006223   \n",
       "\n",
       "          COX1      ATP6        ND3            ND4     TrnL2           ND5  \\\n",
       "0  8639.738516  0.000000   2.317734  362905.328813  0.000000  16425.005888   \n",
       "1  6917.497632  0.000000   2.052061  413379.480243  0.000000   9870.413411   \n",
       "2  6446.875378  0.000000   0.000000  586866.755805  0.000000  12818.421487   \n",
       "3  8238.822253  0.688548  11.361044  145138.020516  0.000000  30920.285112   \n",
       "4  8677.744881  0.000000   0.000000  218419.769390  1.000778  12487.706220   \n",
       "\n",
       "            ND6           CYTB        TrnP  \n",
       "0  25810.282064   66233.102292  505.265938  \n",
       "1  10525.020870   33748.195209  449.401359  \n",
       "2  24771.538933   98703.614050  126.422165  \n",
       "3  34343.401976  138186.438931  104.315037  \n",
       "4  11304.786781   24252.850820  955.742863  \n",
       "\n",
       "[5 rows x 23605 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convert Mouse_ID to string in both dataframes before merging\n",
    "phenotype[\"Mouse_ID\"] = phenotype[\"Mouse_ID\"].astype(str)\n",
    "transcriptome[\"Mouse_ID\"] = transcriptome[\"Mouse_ID\"].astype(str)\n",
    "\n",
    "# Merge phenotype and transcriptome data on Mouse_ID\n",
    "merged = pd.merge(phenotype, transcriptome, on=\"Mouse_ID\")\n",
    "\n",
    "display(merged.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cbf12f7-ca32-47b5-be6b-5f2d76118f05",
   "metadata": {},
   "source": [
    "## Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a7bb3c72-e8ed-43b8-9445-6263fb8ff2ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the phenotype variables to analyze\n",
    "phenotype_vars = args.phenotype_vars\n",
    "\n",
    "# Initialize results dataframe\n",
    "results_df = pd.DataFrame(columns=[\"Phenotype\", \"Permutation Test R² Score\", \"P-value\"])\n",
    "importance_df = pd.DataFrame(columns=[\"Phenotype\", \"Gene\", \"Importance\"])\n",
    "\n",
    "# Define cross-validation strategy\n",
    "cv = KFold(n_splits=5, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6d39033f-5a33-4ad2-8f67-0c4f4f2eac47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Center_occupancy (Adr)...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3903014/1574587672.py:38: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  results_df = pd.concat(\n",
      "/tmp/ipykernel_3903014/1574587672.py:47: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  importance_df = pd.concat(\n",
      "/tmp/ipykernel_3903014/1574587672.py:69: FutureWarning: \n",
      "\n",
      "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.\n",
      "\n",
      "  sns.barplot(data=importance_df.head(10), x=\"Importance\", y=\"Gene\", palette=\"viridis\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Grooming_duration (Adr)...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3903014/1574587672.py:69: FutureWarning: \n",
      "\n",
      "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.\n",
      "\n",
      "  sns.barplot(data=importance_df.head(10), x=\"Importance\", y=\"Gene\", palette=\"viridis\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Social_preference (Adr)...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3903014/1574587672.py:69: FutureWarning: \n",
      "\n",
      "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.\n",
      "\n",
      "  sns.barplot(data=importance_df.head(10), x=\"Importance\", y=\"Gene\", palette=\"viridis\")\n"
     ]
    }
   ],
   "source": [
    "# Loop through each phenotype variable\n",
    "for phenotype_var in phenotype_vars:\n",
    "    print(f\"Processing {phenotype_var} ({args.tissue})...\")\n",
    "\n",
    "    # Define the target variable\n",
    "    y = merged[phenotype_var]\n",
    "\n",
    "    # One-hot encode Treatment as a covariate\n",
    "    covariates = pd.get_dummies(merged[\"Treatment\"], drop_first=True)\n",
    "\n",
    "    covariates = covariates.to_numpy()  # Convert to numpy array\n",
    "    y = y.to_numpy()                    # Ensure y is also a numpy array\n",
    "\n",
    "\n",
    "    # Residualize the phenotype variable against Treatment\n",
    "    cov_model = LinearRegression()\n",
    "    cov_model.fit(covariates, y)\n",
    "    y_pred_cov = cov_model.predict(covariates)\n",
    "    y_res = y - y_pred_cov  # Residualized target variable\n",
    "\n",
    "    # Define transcriptomic features (excluding phenotype metadata)\n",
    "    phenotype_cols = {\"Mouse_ID\", \"Treatment\"} | set(phenotype_vars)\n",
    "    transcriptome_features = [col for col in merged.columns if col not in phenotype_cols]\n",
    "    X_transcriptome = merged[transcriptome_features]\n",
    "\n",
    "    # Define the Random Forest model\n",
    "    rf_model = RandomForestRegressor(n_estimators=20, random_state=42, n_jobs=-1)\n",
    "\n",
    "    # Get cross-validated predictions instead of using training predictions\n",
    "    y_pred_cv = cross_val_predict(rf_model, X_transcriptome, y_res, cv=cv)\n",
    "\n",
    "    # Perform permutation test using cross-validation\n",
    "    score, permutation_scores, pvalue = permutation_test_score(\n",
    "        rf_model, X_transcriptome, y_res, cv=cv, n_permutations=1000, scoring=\"r2\", random_state=42\n",
    "    )\n",
    "\n",
    "    # Store results\n",
    "    results_df = pd.concat(\n",
    "        [results_df, pd.DataFrame({\"Phenotype\": [phenotype_var], \"Permutation Test R² Score\": [score], \"P-value\": [pvalue]})], \n",
    "        ignore_index=True\n",
    "    )\n",
    "\n",
    "    ## Feature Importance\n",
    "    rf_model.fit(X_transcriptome, y_res)\n",
    "\n",
    "    # Store results\n",
    "    importance_df = pd.concat(\n",
    "        [importance_df, pd.DataFrame({\n",
    "        \"Gene\": X_transcriptome.columns,\n",
    "        \"Importance\": rf_model.feature_importances_\n",
    "    }).sort_values(by=\"Importance\", ascending=False)]\n",
    "    )\n",
    "\n",
    "    ### Predicted vs. Actual Plot (Cross-Validation)\n",
    "    slope, intercept, r_value, p_value, std_err = linregress(y_pred_cv, y_res)\n",
    "\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.regplot(x=y_pred_cv, y=y_res, scatter_kws={\"alpha\": 0.6, \"color\": \"blue\"}, line_kws={\"color\": \"red\"})\n",
    "    plt.text(min(y_pred_cv) + (max(y_pred_cv) - min(y_pred_cv)) * 0.05, max(y_res) - (max(y_res) - min(y_res)) * 0.1,\n",
    "             s=f\"p = {p_value:.3g}\", fontsize=8, color=\"black\")\n",
    "    plt.xlabel(f\"Predicted Residualized {phenotype_var} ({args.tissue})\", fontsize=8)\n",
    "    plt.ylabel(f\"Actual Residualized {phenotype_var} ({args.tissue})\", fontsize=8)\n",
    "    plt.title(f\"Predicted vs. Actual - {phenotype_var} ({args.tissue})\", fontsize=8)\n",
    "    plt.savefig(os.path.join(output_dir, f\"Predicted_vs_actual_{phenotype_var}_{args.tissue}.pdf\"), format=\"pdf\")\n",
    "    plt.show()\n",
    "\n",
    "    ### Feature Importance Plot\n",
    "    plt.figure(figsize=(6, 4))\n",
    "    sns.barplot(data=importance_df.head(10), x=\"Importance\", y=\"Gene\", palette=\"viridis\")\n",
    "    plt.xlabel(\"Feature Importance\")\n",
    "    plt.ylabel(\"Genes\")\n",
    "    plt.title(f\"Top 10 Important Genes for Prediction ({args.tissue})\")\n",
    "    plt.savefig(os.path.join(output_dir, f\"Feature_importance_{phenotype_var}_{args.tissue}.pdf\"), format=\"pdf\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0b36d426-c0cf-4810-9e4f-f6b3391974e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save final results to CSV\n",
    "results_df.to_csv(os.path.join(output_dir, f\"RF_permutation_test_results_{args.tissue}.csv\"), index=False)\n",
    "importance_df.to_csv(os.path.join(output_dir, f\"RF_important_features_{args.tissue}.csv\"), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
